### PR TITLE
WIP: Detection re-runs in RUNNING DetectionStatus #699

### LINKED ISSUE #699: Detection re-runs in RUNNING DetectionStatus
If timerange is in `RUNNING` DetectionStatus, detection re-runs on `/span` query. Thus, you can get infinite `RUNNING` status.
It should just return `RUNNING` instead.
P.S. maybe the reason is span extending?

### PR DESCRIPTION
**Pull request by amper43**
fixes #699 

## Changes
- filtering spans that inside or intersected with `FAILED` spans

### CODE REVIEW

File: server/src/controllers/analytics_controller.ts [hunk 1]
```diff
@@ -560,16 +561,35 @@ export async function getDetectionSpans(
...
+    if(!_.isEmpty(insideFailed)) {
+      return;
+    }
+
+    const intersectedFailedSpans = await Detection.getIntersectedSpans(analyticUnitId, from, to, Detection.DetectionStatus.FAILED);
+    const intersectedFailedSpansBorders = Detection.getSpanBorders(intersectedFailedSpans);
+    const filteredSpans = getNonIntersectedSpans(from, to, intersectedFailedSpansBorders);
+    _.concat(filteredNewDetectionSpans, filteredSpans);
+  });
+  Promise.all(filterFailedSpanPromises);
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  await Promise.all(filterFailedSpanPromises);
```

File: server/src/controllers/analytics_controller.ts [hunk 2]
```diff
@@ -560,16 +561,35 @@ export async function getDetectionSpans(
...
+    const intersectedFailedSpansBorders = Detection.getSpanBorders(intersectedFailedSpans);
+    const filteredSpans = getNonIntersectedSpans(from, to, intersectedFailedSpansBorders);
+    _.concat(filteredNewDetectionSpans, filteredSpans);
+  });
+  await Promise.all(filterFailedSpanPromises);
+
   let runningSpansPromises = [];
   let newRunningSpans: Detection.DetectionSpan[] = [];
-  runningSpansPromises = newDetectionSpans.map(async span => {
+  runningSpansPromises = filteredNewDetectionSpans.map(async span => {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  const runningSpansPromises = filteredNewDetectionSpans.map(async span => {
```

File: server/src/controllers/analytics_controller.ts [hunk 3]
```diff
@@ -560,16 +561,35 @@ export async function getDetectionSpans(
...
+    }
+
+    const intersectedFailedSpans = await Detection.getIntersectedSpans(analyticUnitId, from, to, Detection.DetectionStatus.FAILED);
+    const intersectedFailedSpansBorders = Detection.getSpanBorders(intersectedFailedSpans);
+    const filteredSpans = getNonIntersectedSpans(from, to, intersectedFailedSpansBorders);
+    _.concat(filteredNewDetectionSpans, filteredSpans);
+  });
+  await Promise.all(filterFailedSpanPromises);
+
   let runningSpansPromises = [];
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
```

**Comment by amper43** [comment]
Actual state:
Changes of span's state machine is need. Current implementation of PR works incorrect in case when panel requests range, that contain failed span. After that unit become to `FAILED` sate and can't run detections.

**Pull request closed without merging**